zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/MessageController.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.gateway.rest.controller;
+
+import io.camunda.service.MessageServices;
+import io.camunda.service.MessageServices.CorrelateMessageRequest;
+import io.camunda.zeebe.gateway.impl.configuration.MultiTenancyCfg;
+import io.camunda.zeebe.gateway.protocol.rest.MessageCorrelationRequest;
+import io.camunda.zeebe.gateway.protocol.rest.MessageCorrelationResponse;
+import io.camunda.zeebe.gateway.rest.RequestMapper;
+import io.camunda.zeebe.gateway.rest.ResponseMapper;
+import io.camunda.zeebe.gateway.rest.RestErrorMapper;
+import java.util.concurrent.CompletableFuture;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@CamundaRestController
+@RequestMapping("/v2/messages")
+public class MessageController {
+
+  private final MessageServices<MessageCorrelationResponse> messageServices;
+  private final MultiTenancyCfg multiTenancyCfg;
+
+  @Autowired
+  public MessageController(
+      final MessageServices<MessageCorrelationResponse> messageServices,
+      final MultiTenancyCfg multiTenancyCfg) {
+    this.messageServices = messageServices;
+    this.multiTenancyCfg = multiTenancyCfg;
+  }
+
+  @PostMapping(
+      path = "/correlation",
+      produces = {MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_PROBLEM_JSON_VALUE},
+      consumes = MediaType.APPLICATION_JSON_VALUE)
+  public CompletableFuture<ResponseEntity<MessageCorrelationResponse>> correlateMessage(
+      @RequestBody final MessageCorrelationRequest correlationRequest) {
+    return RequestMapper.toMessageCorrelationRequest(
+            correlationRequest, multiTenancyCfg.isEnabled())
+        .fold(this::correlateMessage, RestErrorMapper::mapProblemToCompletedResponse);
```

Comment:
🔧 If I am not mistaken, the `fold` message has already been adjusted to swap left and right functions (left going, well, left now 🙂).